docs/advanced-guides/testing/testing-with-react-testing-library.md

```diff
@@ -1,13 +1,153 @@
-# Testing With React Testing Library
+# Testing With React Testing Library (RTL)
+
+`react-testing-library` is a lightweight family of packages that help us test UI components in a manner that resembles the way users interact with our applications.
+
+To quote their (very excellent) documentation:
+
+    The more your tests resemble the way your software is used, the more confidence they can give you.
+
+https://testing-library.com/docs/intro
 
 ## Getting Setup
 
-TODO
+This guide assumes you followed the instructions for [Adding React Router via Create React App](../../installation/add-to-a-website.md#create-react-app). If you did not start with Create React App, but still have the same application structure, you can install [React Testing Library](https://testing-library.com/docs/react-testing-library/intro) by following [their installation guide](https://github.com/testing-library/react-testing-library#installation).
 
-## Testing Routes and Redirects
+## Basic Test
+
+A basic rendering test ensures that we have everything installed and set up correctly. Fortunately, RTL gives us the tools to accomplish this.
 
-TODO
+Since we've wrapped our `App` component in the `Router` in our `index.js` file, we do have to wrap it around each of our individual component tests, otherwise `history` will be undefined. If the `Router` had been inside of our `App`, then we would not have to wrap it inside of our tests.
+
+A recommended test to ensure basic functionality looks like the following:
+
+```jsx
+test('<App /> renders successfully', () => {
+  render(<App />, { wrapper: Router });
+  const header = screen.getByText('Welcome to React Router!');
+  expect(header).toBeInTheDocument();
+});
+```
+
+This test ensures that the `App` component renders, and that the `h1` we added during the setup guide exists in the document.
 
 ## Testing Links and Navigation
 
-TODO
+`react-router` has a lot of tests verifying that the routes work when the location changes, so you probably don't need to test this stuff.
+
+If you need to test navigation within your app, you can do so by firing a click event on the link itself and asserting that the path changed to the expected value.
+
+This is accomplished like so:
+
+```jsx
+it('navigates to /about', () => {
+  render(<App />, { wrapper: Router });
+
+  const aboutLink = screen.getByText('About');
+  fireEvent.click(aboutLink);
+
+  expect(window.location.pathname).toBe('/about');
+});
+```
+
+Since our setup guide then replaces the about link with a link back to the home page, we can test navigating back to the home page by triggering another click, this time on the home link:
+
+```jsx
+it('navigates to /about and back to /', () => {
+  render(<App />, { wrapper: Router });
+
+  const aboutLink = screen.getByText('About');
+  fireEvent.click(aboutLink);
+
+  expect(window.location.pathname).toBe('/about');
```

Comment:
This won't be the case if you use an in-memory router (it does not update the window.location). I think there are different opinions about this. The downside of using a "full" router is that JSDOM doesn't give you a way to reset the navigation stack, so if you do something like "go back 3 times" you could end up on a previous test's URL. On the other hand, memory router doesn't represent what's going in the actual app as well, especially if you have some regular `<a>` tags mixed in or manual manipulation of the URL.

Reply:
I'm using MemoryRouter in my tests. However, I do find that I need to define the router once per test, otherwise the route history doesn't reset. (Would love to know how we could reset the navigation state before tests)

```
 test("Should redirect to the home screen if 'action' query param is missing", async () => {
    function HomePage() {
      return <h1>Home</h1>;
    }
    const { findByRole } = render(
        <Router initialEntries={["/redirect"]}>
          <Routes>
            <Route path="/" element={<HomePage />} />
            <Route path="/redirect" element={<RedirectPage />} />
          </Routes>
        </Router>
    );
    const homePageHeading = await findByRole(/heading/i);
    expect(homePageHeading).toBeVisible();
  });
```

Reply:
Yes, if you test the page contents, you can make an assertion. But you can't assert `window.location.pathname` with MemoryRouter, because it doesn't update that.

Reply:
This is why I've moved away from using the MemoryRouter and instead I render the *same* router that's rendered in my source code and interact with window.location directly instead: https://github.com/kentcdodds/bookshelf/blob/d9d70fceaea790dfe57ee5a46c13f9f9cc810675/src/test/app-test-utils.js#L15

Works *way* better. **And** I get more confidence out of my tests.